File: prototype/rails/src/app/domain/notifications/notification_message.rb
require_relative "../money"

module Domain
  module Notifications
    NotificationMessage = Data.define(:subject, :body, :url) do
      def self.item_sold(order_id, net)
        new(
          subject: "Item sold",
          body: "Order ##{order_id} is paid. #{net.format} is held until the customer confirms delivery.",
          url: nil
        )
      end

      def self.order_shipped(order_id, carrier, tracking_number)
        new(
          subject: "Order shipped",
          body: "Order ##{order_id} shipped with #{carrier}. Tracking number #{tracking_number}.",
          url: nil
        )
      end
    end
  end
end

File: prototype/php/src/app/Domain/Notifications/NotificationMessage.php
<?php

declare(strict_types=1);

namespace App\Domain\Notifications;

use App\Domain\Money\Money;

final readonly class NotificationMessage
{
    private function __construct(public string $subject, public string $body, public ?string $url) {}

    public static function itemSold(string $orderId, Money $net): self
    {
        return new self(
            'Item sold',
            "Order {$orderId} is paid. {$net->format()} is held until the customer confirms delivery.",
            null,
        );
    }

    public static function orderShipped(string $orderId, string $carrier, string $trackingNumber): self
    {
        return new self(
            'Order shipped',
            "Order {$orderId} shipped with {$carrier}. Tracking number {$trackingNumber}.",
            null,
        );
    }

    /**
     * $url names the recipient's thread on their own site. Nullable for the
     * same reason every factory here is: an inbox row can carry no link.
     */
    public static function messageReceived(string $topic, ?string $url): self
    {
        return new self(
            'New message',
            "You have a new message about {$topic}.",
            $url,
        );
    }

    /**
     * The shape an inbox row stores and a page reads back.
     *
     * @return array{subject: string, body: string, url: string|null}
     */
    public function toArray(): array
    {
        return [
            'subject' => $this->subject,
            'body' => $this->body,
            'url' => $this->url,
        ];
    }
}
